zcash_proofs/src/lib.rs

```diff
@@ -60,58 +88,196 @@ pub fn default_params_folder() -> Option<PathBuf> {
+/// Always checks the size and hash of the file, even if it didn't need to be downloaded.
+///
+/// See [`download_sapling_parameters`] for details.
+#[cfg(feature = "download-params")]
+#[cfg_attr(docsrs, doc(cfg(feature = "download-params")))]
+fn fetch_params(
+    name: &str,
+    expected_hash: &str,
+    expected_bytes: u64,
+    timeout: Option<u64>,
+) -> Result<PathBuf, minreq::Error> {
     // Ensure that the default Zcash parameters location exists.
     let params_dir = default_params_folder().ok_or_else(|| {
         io::Error::new(io::ErrorKind::Other, "Could not load default params folder")
     })?;
     std::fs::create_dir_all(&params_dir)?;
 
-    let fetch_params = |name: &str, expected_hash: &str| -> Result<(), minreq::Error> {
-        use std::io::Write;
-
-        // Download the parts directly (Sapling parameters are small enough for this).
-        let part_1 = minreq::get(format!("{}/{}.part.1", DOWNLOAD_URL, name)).send()?;
-        let part_2 = minreq::get(format!("{}/{}.part.2", DOWNLOAD_URL, name)).send()?;
-
-        // Verify parameter file hash.
-        let hash = blake2b_simd::State::new()
-            .update(part_1.as_bytes())
-            .update(part_2.as_bytes())
-            .finalize()
-            .to_hex();
-        if &hash != expected_hash {
-            return Err(io::Error::new(
-                io::ErrorKind::InvalidData,
-                format!(
-                    "{} failed validation (expected: {}, actual: {}, fetched {} bytes)",
-                    name,
-                    expected_hash,
-                    hash,
-                    part_1.as_bytes().len() + part_2.as_bytes().len()
-                ),
-            )
-            .into());
+    let params_path = params_dir.join(name);
+
+    // Download parameters if needed.
+    // TODO: use try_exists when it stabilises, to exit early on permissions errors (#83186)
+    if !params_path.exists() {
+        let result = stream_params_downloads_to_disk(
+            &params_path,
+            name,
+            expected_hash,
+            expected_bytes,
+            timeout,
+        );
+
+        // Remove the file on error, and return the download or hash error.
+        if result.is_err() {
+            let _ = std::fs::remove_file(&params_path);
+            result?;
         }
+    } else {
+        // TODO: avoid reading the files twice
+        // Either:
+        // - return Ok if the paths exist, or
+        // - always load and return the parameters, for newly downloaded and existing files.
+
+        let file_path_string = params_path.to_string_lossy();
+
+        // Check the file size is correct before hashing large amounts of data.
+        verify_file_size(&params_path, expected_bytes, name, &file_path_string).expect(
+            "parameter file size is not correct, \
+             please clean your Zcash parameters directory and re-run `fetch-params`.",
```

Comment:
Should this error reference `fetch-params` or the `download-params` tool in this repo?

Reply:
I wrote something similar to the previous `panic!()` text, but I'm happy to change it if you'd like:
> panic!("Sapling spend parameter file is not correct, please clean your `~/.zcash-params/` and re-run `fetch-params`.");

https://github.com/zcash/librustzcash/pull/459/files#diff-67a539a91d2370483ef0aef0e6d89f1d1b480eaf9767975709ff767ec3a282b8L182